crates/matrix-sdk/tests/integration/send_queue.rs

```diff
@@ -2085,3 +2079,86 @@ async fn test_unwedging_media_upload() {
     // That's all, folks!
     assert!(watch.is_empty());
 }
+
+#[async_test]
+async fn test_media_upload_ordering() {
+    // Test that despite happening in multiple requests, sending a media maintains
+    // the ordering.
+
+    let mock = MatrixMockServer::new().await;
+
+    // Mark the room as joined.
+    let room_id = room_id!("!a:b.c");
+    let client = mock.client_builder().build().await;
+    let room = mock.sync_joined_room(&client, room_id).await;
+
+    let q = room.send_queue();
+    let (local_echoes, mut watch) = q.subscribe().await.unwrap();
+    assert!(local_echoes.is_empty());
+
+    // Prepare endpoints.
+    mock.mock_room_state_encryption().plain().mount().await;
+    mock.mock_upload().ok(mxc_uri!("mxc://sdk.rs/media")).mock_once().mount().await;
+
+    assert!(watch.is_empty());
+
+    {
+        // 1. Send a text message that will get wedged.
+        mock.mock_room_send().error_too_large().mock_once().mount().await;
+        q.send(RoomMessageEventContent::text_plain("error").into()).await.unwrap();
+        let (text_txn, _send_handle) = assert_update!(watch => local echo { body = "error" });
+        assert_update!(watch => error { recoverable = false, txn = text_txn });
+    }
+
+    // We'll then send a media event, and then a text event with success.
+    mock.mock_room_send().ok(event_id!("$media")).mock_once().mount().await;
+    mock.mock_room_send().ok(event_id!("$text")).mock_once().mount().await;
+
+    // 2. Queue the media.
+    let filename = queue_attachment_no_thumbnail(&q).await;
+
+    // 3. Queue the message.
+    q.send(RoomMessageEventContent::text_plain("hello world").into()).await.unwrap();
+
+    // Observe the local echo for the media.
+    let (event_txn, _send_handle, content) = assert_update!(watch => local echo event);
+    assert_let!(MessageType::Image(img_content) = content.msgtype);
+    assert_eq!(img_content.body, filename);
+
+    // Observe the local echo for the message.
+    let (text_txn, _send_handle) = assert_update!(watch => local echo { body = "hello world" });
+
+    // The media gets uploaded.
+    assert_update!(watch => uploaded {
+        related_to = event_txn,
+        mxc = mxc_uri!("mxc://sdk.rs/media")
+    });
+
+    let edit_msg = assert_update!(watch => edit local echo {
```

Comment:
Doesn't have to be a separate test; this happens for all media uploads: as soon as we get the final MXC IDs for the media, the media event which used local MXC IDs sees its IDs replaced with the final one, which is modeled by an "edit" of the local echo, just before the media event is actually sent to the server.

I'll add a tiny comment; this is mostly a sanity check at this point, since this behavior is well tested in other places.